src/rgw/rgw_rados.cc

```diff
@@ -7260,6 +7260,65 @@ int RGWRados::bucket_index_read_olh_log(const RGWBucketInfo& bucket_info, RGWObj
   return 0;
 }
 
+// a multisite sync bug resulted in the OLH head attributes being overwritten by
+// the attributes from another zone, causing link_olh() to fail endlessly due to
+// olh_tag mismatch. this attempts to detect this case and reconstruct the OLH
+// attributes from the bucket index. see http://tracker.ceph.com/issues/37792
+int RGWRados::repair_olh(RGWObjectCtx& obj_ctx, RGWObjState *state,
+                         const RGWBucketInfo& bucket_info, const rgw_obj& obj,
+                         rgw_zone_set *zones_trace)
+{
+  // fetch the current olh entry from the bucket index
+  rgw_bucket_olh_entry olh;
+  int r = bi_get_olh(bucket_info, obj, &olh);
+  if (r < 0) {
+    ldout(cct, 0) << "repair_olh failed to read olh entry for " << obj << dendl;
+    return r;
+  }
+  if (olh.tag == state->olh_tag.to_str()) { // mismatch already resolved?
```

Comment:
@cbodley we were pretty inconsistent in the way we stored object attrs. They can be either null terminated or not. `bufferlist.to_str()` will return a different string in both cases (the null character at the end will be part of the string), and this check could fail. Need to make sure that `state->olh_tag` is sanitized.

Reply:
the olh_tag bufferlist is stored without the null. RGWRados::olh_init_modification_impl() generates them with `olh_bl.append(olh_tag.c_str(), olh_tag.size());`

it looks like we have a rgw_bl_str() helper function that trims trailing nulls though, i can use that to be safe